delivery_carrier_label_batch/wizard/generate_labels.py

```diff
@@ -0,0 +1,280 @@
+        error_queue = queue.Queue()
+
+        # If we have more than one pack in a picking, we must ensure
+        # they are not executed concurrently or we will have concurrent
+        # transaction errors. So we process them in the same thread.
+        # We put them in the same 'group' and this group will be passed
+        # as a whole to a thread worker.
+        groups = {}
+        for pack, operations, label in self._get_packs(batch):
+            if not label or self.generate_new_labels:
+                picking = operations[0].picking_id
+                groups.setdefault(picking.id, []).append(
+                    (pack, picking, label)
+                )
+
+        for group in groups.values():
+            data_queue.put(group)
+
+        # create few workers to parallelize label generation
+        num_workers = self._get_num_workers()
+        _logger.info('Starting %s workers to generate labels', num_workers)
+        for i in range(num_workers):
+            t = threading.Thread(target=self._worker,
+                                 args=(data_queue, error_queue))
+            t.daemon = True
+            t.start()
+
+        # wait for all tasks to be done
+        data_queue.join()
+
+        # We will not create a partial PDF if some labels weren't
+        # generated thus we raise catched exceptions by the workers
+        # We will try to regroup all orm exception in one
+        if not error_queue.empty():
+
+            error_count = {}
+            messages = []
+            while not error_queue.empty():
+                e = error_queue.get()
+                if isinstance(e, exceptions.UserError):
+                    if e.name not in error_count:
+                        error_count[e.name] = 1
+                    else:
+                        error_count[e.name] += 1
+                    messages.append(str(e) or '')
+                else:
+                    # raise other exceptions like PoolError if
+                    # too many cursor where created by workers
+                    raise e
+            titles = []
+            for key, v in error_count.items():
+                titles.append('%sx %s' % (v, key))
+
+            message = _('Some labels couldn\'t be generated. Please correct '
+                        'following errors and generate labels again to create '
+                        'the ones which failed.\n\n'
+                        ) + '\n'.join(messages)
+            raise exceptions.UserError(message)
+
+        # create a new cursor to be up to date with what was created by workers
+        with self._do_in_new_env() as new_env:
+            # labels = new_env['shipping.label']
+            self_env = self.with_env(new_env)
+            labels = []
+            for pack, operations, label in self_env._get_packs(batch):
+                picking = operations[0].picking_id
+                if pack:
+                    label = self_env._find_pack_label(pack)
+                else:
```

Comment:
```suggestion
                elif not label and picking:
```
@sebalix should this condition check for both cases?

Reply:
IIRC we wants to enforce the use of packs and not try to print the labels if the packs are missings.

Testing locally having pickings without packs it wasn't working.